File: d23.hs
{-# LANGUAGE ForeignFunctionInterface #-}

import           Data.Char                      ( isSpace
                                                , ord
                                                )
import           Data.List                      ( elemIndex )
import           Data.Maybe                     ( fromJust )
import qualified Data.Vector.Storable          as SV
import           Foreign                        ( )
import           Foreign.Ptr
import           Foreign.C.Types
import           System.IO.Unsafe               ( unsafePerformIO )
import           Paths_aoc20_hs                 ( getDataFileName )

main :: IO ()
main = do
  fname <- getDataFileName "d23.txt"
  input <- readFile fname
  let initialState = map (\c -> ord c - ord '0') . filter (not . isSpace) $ input

  putStr "Part 1 (with Data.Sequence): "
  putStrLn $ getOrder . (!! 100) $ iterate update initialState

  -- runGame doesn't return the entire game state, only the product of the two
  -- cups after cup 1 (i.e. the requested solution for Part 2).
  putStr "Part 1 product of two cups after cup 1 (with FFI): "
  print $ runGame initialState 0 100

  putStr "Part 2 (with FFI): "
  print $ runGame initialState 1000000 10000000

update :: [Int] -> [Int]
update cups = cups'
  where
    currentCup:rest = cups
    (pickedUpCups, otherCups) = splitAt 3 rest
    destinationCup = case filter (< currentCup) otherCups of
      [] -> maximum otherCups
      xs -> maximum xs
    destinationIndex = fromJust . elemIndex destinationCup $ otherCups
    (other1, other2) = splitAt (destinationIndex + 1) otherCups
    -- We place the currentCup at the back so that the new head of the list
    -- will be the currentCup for the next round.
    cups' = mconcat [other1, pickedUpCups, other2, [currentCup]]

getOrder :: [Int] -> String
getOrder cups = ans
  where
    index1 = fromJust $ elemIndex 1 cups
    (before1, _ : after1) = splitAt index1 cups
    ans = concatMap show (after1 ++ before1)

-- The C function signature:
-- long int run_game(int *first_input, int first_input_len, int full_len, int nrounds);
foreign import ccall unsafe "d23.h run_game"
  c_runGame :: Ptr CInt      -- The initial values specified in the puzzle input.
               -> CInt       -- The length of those initial values.
               -> CInt       -- The value to pad up until. Must be at least equal to
                             -- the length of initial values. This is not checked in C
                             -- code, but the Haskell wrapper does check.
               -> CInt       -- The number of rounds to play.
               -> IO CLong   -- The final score.

-- And the Haskell function which uses it:
runGame :: [Int]      -- The initial values. We don't need to provide a length.
           -> Int     -- The number to pad up until.
           -> Int     -- The number of rounds to play.
           -> Integer -- The final score.
runGame initial padding rounds = score
  where
    initialSV = SV.fromList . map fromIntegral $ initial
    len = SV.length initialSV
    len' = fromIntegral len
    padding' = fromIntegral (max padding len) -- make sure that padding' >= len.
    rounds' = fromIntegral rounds
    score = fromIntegral . unsafePerformIO $
      SV.unsafeWith initialSV (\ptr -> c_runGame ptr len' padding' rounds')

File: d23.c
#include <stdio.h>
#include <stdlib.h>
#include "d23.h"

/* int main (void) */
/* { */
/*     int first_input[] = {6, 2, 4, 3, 9, 7, 1, 5, 8}; */
/*     long int ans = run_game(first_input, 9, 1000000, 10000000); */
/*     printf("%ld\n", ans); */
/* } */

long int run_game(int *first_input, int first_input_len, int full_len, int nrounds) {
    // Initialise the game array
    int *game = malloc(sizeof(int) * full_len + 1);
    // Fill in the provided values. game[n] will contain the number that is after n.
    game[0] = 0;  // Helps me avoid off-by-one errors.
    for (int i = 0; i < first_input_len - 1; i++) {
        game[first_input[i]] = first_input[i + 1];
    }
    // Pad with extra numbers if necessary.
    if (full_len > first_input_len) {
        game[first_input[first_input_len - 1]] = first_input_len + 1;
        for (int i = first_input_len + 1; i < full_len; i++) {
            game[i] = i + 1;
        }
        game[full_len] = first_input[0];
    }
    // Otherwise, just close the loop.
    else {
        game[first_input[first_input_len - 1]] = first_input[0];
    }
    int current = first_input[0];

    for (int i = 0; i < nrounds; i++) {
        game = update(game, &current, full_len);
    }
    long int score = count_score(game);
    free(game);
    return score;
}


int *update(int *game, int *current, int length)
{
    // Pick out the three cups
    int a = game[*current];
    int b = game[a];
    int c = game[b];
    int d = game[c];
    game[*current] = d;

    // Figure out what the destination cup is. It needs to be smaller than the
    // current value, *unless* there is no such cup in which case it needs to
    // be the largest available cup.
    int destination = *current - 1;
    for (;;) {
        if (destination < 1) destination = length;
        if (destination == a || destination == b || destination == c) {
            destination--;  // was picked up already
        }
        else break;
    }

    // Insert the three cups back in
    game[c] = game[destination]; // link third cup -> cup after destination
    game[destination] = a;       // link destination -> first cup
    // Move the current cup to the next
    *current = d;
    
    return game;
}

void print_game(int *game, int *current, int length)
{
    int a = *current;
    for (int i = 0; i < length; i++) {
        printf("%d ", a);
        a = game[a];
    }
    printf("\n");
}

long int count_score(int *game)
{
    return (long int) game[1] * game[game[1]];
}
